gymnasium/functional.py

```diff
@@ -0,0 +1,86 @@
+"""Base class and definitions for an alternative, functional backend for gym envs, particularly suitable for hardware accelerated and otherwise transformed environments."""
+
+from typing import Any, Callable, Dict, Generic, Optional, Tuple, TypeVar
+
+import numpy as np
+
+StateType = TypeVar("StateType")
+ActType = TypeVar("ActType")
+ObsType = TypeVar("ObsType")
+RewardType = TypeVar("RewardType")
+TerminalType = TypeVar("TerminalType")
+RenderStateType = TypeVar("RenderStateType")
+
+
+class FuncEnv(Generic[StateType, ObsType, ActType, RewardType, TerminalType]):
```

Comment:
Add `RenderStateType` and add `abc.ABC`?

Reply:
Added RenderStateType, I think I'll keep ABC out for now, I'd need to triple check if it wouldn't break anything (especially with jax stuff). It shouldn't, but...